openedx/openedx-platform
# ENT-839 Add ability to configure SP metadata for an IdP

This PR introduces the ability to configure different saml service provider metadata by identity provider.
The SAMLConfiguration model now has a slug field that is keyed with site - existing configurations will become the "default". This means in any context where no identity provider is specified, this default configuration will be used, as it is today.
The ability to read the x509 cert and secret key from the settings file is still preserved. For all default configurations (there is only one of these per site), it will read from the existing settings, for other named configurations it will read from a new setting that is a dictionary of the slug name to the cert/key.
The SAMLProviderConfig model now has an optional foreign key to the SAMLConfiguration. If it is not set, it will use the default one for the site. If it is set, the cert/key used to verify the SAML handshake will come from the SAMLConfiguration it points to.
This means that when we give our metadata to customers, if we need to access a particular version of the SAMLConfiguration, we need only append tpa_hint=<idp_slug> as a query parameter to the metadata.xml url and the correct cert will be returned.

Testing Instructions:
Verify that existing SAMLConfigurations have the slug 'default'.
Create a new SAMLConfiguration with a different slug name. You can either save the cert and key values in the model, or edit /edx/app/edxapp/lms.auth.json to add the following settings:

`"SOCIAL_AUTH_SAML_SP_PRIVATE_KEY_DICT": {
        "slug_name": "<private_key>"
    },
    "SOCIAL_AUTH_SAML_SP_PUBLIC_CERT_DICT": {
        "slug_name": "<public_cert>"
    }`

Hit the metadata endpoint for the server, ie https://brittneyexline.sandbox.edx.org/auth/saml/metadata.xml

Decide on one of the configured SamlProviderConfigs that you plan to test with, ensure that it does not have saml_configuration set, and hit the same endpoint with a tpa parameter, ie https://brittneyexline.sandbox.edx.org/auth/saml/metadata.xml?tpa_hint=testshib
At this point, the xml should be the same as the endpoint without the parameter.

Go to the SamlProviderConfig and change the saml_configuration to the one that you created, and save. Visit the endpoint again, it should be returning the public cert configured from the SAMLConfiguration you added.

Now, create a file (with the name `my_sandbox_metadata_b30a4d20-0fbe-4c0b-9719-f276bb4aa035.xml` for the brittneyexline sandbox) and paste the metadata from the tpa_hinted endpoint. Upload it to testshib. Go to login page and click on the testshib provider icon (configured to show on the brittneyexline sandbox). SSO should be triggered without error.

Now switch the SAMLProviderConfig back to using the default SAMLConfiguration and/or unset that field and save. Copy the metadata from the same endpoint, it should be back to the original, and save it in the metadata file. Upload this file to testshib and login using testshib SSO, it should be successful.
